packages/devtools_app/lib/src/timeline/timeline_controller.dart

```diff
@@ -249,22 +193,163 @@ class TimelineController {
     }
   }
 
-  Future<void> exitOfflineMode({bool clearTimeline = true}) async {
+  void exitOfflineMode({bool clearTimeline = true}) {
     if (clearTimeline) {
-      await timelineData.clear();
+      frameBasedTimeline.data.clear();
     }
     offlineTimelineData = null;
   }
 
-  Future<double> get displayRefreshRate async =>
-      timelineData?.displayRefreshRate ??
-      await serviceManager.getDisplayRefreshRate();
+  Future<void> clearData() async {
+    if (serviceManager.hasConnection) {
+      await serviceManager.service.clearVMTimeline();
+    }
+    frameBasedTimeline.clear();
+    fullTimeline.clear();
+    allTraceEvents.clear();
+  }
+
+  void recordTrace(Map<String, dynamic> trace) {
+    final traceEvents = timelineMode == TimelineMode.frameBased
+        ? frameBasedTimeline.data.traceEvents
```

Comment:
this another case where having a currentTimeline getter would simplify things a little.